helix-term/src/commands.rs

```diff
@@ -1145,9 +1155,16 @@ fn search_impl(doc: &mut Document, view: &mut View, contents: &str, regex: &Rege
 
     // use find_at to find the next match after the cursor, loop around the end
     // Careful, `Regex` uses `bytes` as offsets, not character indices!
-    let mat = regex
-        .find_at(contents, start)
-        .or_else(|| regex.find(contents));
+    let mat = match direction {
+        Direction::Forward => regex
+            .find_at(contents, start)
+            .or_else(|| regex.find(contents)),
+        Direction::Backward => regex
+            .find_iter(&contents[..start])
+            .fold((None, None), |(_x, y), new| (y, Some(new)))
+            .0
```

Comment:
Shouldn't this just be `.last()`? Right now it seems to return the occurrence just before the last occurrence.

Reply:
We should determine `start` differently for reverse search:

```rust
     let start = text.char_to_byte(graphemes::prev_grapheme_boundary(
        text,
        selection.primary().from(), // this gets the left edge of the selection
    ));
```

That seems to match kakoune too: `[this] word t`, reverse searching for `t` will find `this word [t]`, skipping the start of `this`.

We likely also want to use `.to()` for forward search instead of `.cursor()`

With this change you should be able to use `.last()`

Reply:
`prev_grapheme_boundary`, awesome, that'll simplify that a lot.

Reply:
Hm. If we set start to be before beginning of the selection, than forward search hits the selection itself as a match. If we set start to be after the beginning of the selection, than backward search hits the selection (at least, with a 1-character selection). All matching on before the left instead of after the right does is make it so that forward search needs to get the second, instead of backwards getting the second last.

Reply:
Yeah, you need to determine `start` differently based on the direction:

```rust
     let start = match direction {
         Direction::Forward => {
              text.char_to_byte(graphemes::next_grapheme_boundary(
                  text,
                  selection.primary().to(), // this gets the right edge of the selection
              ))
         }
         Direction::Backward => {
              text.char_to_byte(graphemes::prev_grapheme_boundary(
                  text,
                  selection.primary().from(), // this gets the left edge of the selection
              ))
         }
    }
```